SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Utils/Synchronization/StatefulMutex.cs

```diff
@@ -56,31 +114,45 @@ public void Lock(State stateSnapshot)
         public async Task LockAsync(State stateSnapshot)
         {
             await _mutex.WaitAsync().ConfigureAwait(false);
-            if (!_stateKeeper.IsCurrent(stateSnapshot))
+            if (_isShutdown)
+            {
+                throw new StatefulMutexException(ShutdownExceptionMessage);
+            }
+            if (!IsCurrent(stateSnapshot))
             {
                 Unlock();
-                throw new StatefulMutexException("Cannot lock mutex with expired state");
+                throw new StatefulMutexException(StateExceptionMessage);
             }
         }
 
         /// <summary>
-        /// Asynchronously locks the mutex and does not verify any state
-        /// </summary>
-        /// <seealso cref="Lock"/>
-        public Task LockAsync()
-        {
-            return _mutex.WaitAsync();
-        }
-
-        /// <summary>
         /// Unlocks the mutex. Consecutive calls will not throw an exception
         /// </summary>
         public void Unlock()
         {
-            if (_mutex.CurrentCount == 0)
+            if (_mutex.CurrentCount != 0)
             {
-                _mutex.Release();
+                throw new StatefulMutexException("Trying to unlock not locked mutex");
             }
+            _mutex.Release();
+        }
+
+        public LockHolder GetLock()
+        {
+            Lock();
+            return new LockHolder(this);
+        }
+
+        public async Task<LockHolder> GetLockAsync()
```

Comment:
Why not just make Lock() return a LockHolder (and same for async versions)? That way we don't have more methods than needed